Mathlib/Set.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2016 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Leonardo de Moura
+-/
+import Mathlib.Mem
+/-!
+
+# Sets
+
+This file sets up the theory of subsets of a type.
+
+## Main definitions
+
+* `{a : X | p a} : Set X` : the subset of `X` cut out by the predicate `p : X → Prop`
+* `{a | p a} : Set X` : equal to `{a : X | p a}`
+* `{a ∈ S | p a} : Set X` : the subset of `X` consisting of the elements of `S : set X`
+     which are cut out by `p`.
+
+## Implementation issues
+
+As in Lean 3, `Set X := X → Prop`
+
+I didn't call this file Data.Set.Basic because it contains core Lean 3
+stuff which happens before mathlib3's data.set.basic .
+This file is a port of the core Lean 3 file `lib/lean/library/init/data/set.lean`.
+
+## TODO
+
+Notation {a,b,c} for finite sets (both parser and prettyprinter).
+
+-/
+
+universes u v
+
+def Set (α : Type u) := α → Prop
+
+def setOf {α : Type u} (p : α → Prop) : Set α :=
+p
+
+class Subset (α : Type u) :=
+(subset : α → α → Prop)
```

Comment:
@gebner should these classes (`Subset`, `Union`, `Inter`, `Sdiff`) also get "where"d?

Reply:
I'd prefer it.

Reply:
Done